proto_parser/src/proto_value.h

```diff
@@ -0,0 +1,88 @@
+/*
+*   Copyright 2020 Google LLC
+*
+*   Licensed under the Apache License, Version 2.0 (the "License");
+*   you may not use this file except in compliance with the License.
+*   You may obtain a copy of the License at
+*
+*    https://www.apache.org/licenses/LICENSE-2.0
+*
+*   Unless required by applicable law or agreed to in writing, software
+*   distributed under the License is distributed on an "AS IS" BASIS,
+*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*   See the License for the specific language governing permissions and
+*   limitations under the License.
+*/
+
+#ifndef WIRELESS_ANDROID_PLAY_ANALYTICS_PROTO_PARSER_SRC_PROTO_VALUE_H
+#define WIRELESS_ANDROID_PLAY_ANALYTICS_PROTO_PARSER_SRC_PROTO_VALUE_H
+
+#include <gtest/gtest.h>
+#include <assert.h>
+#include <google/protobuf/text_format.h>
+#include <vector>
+#include <memory>
+#include <sstream>
+
+#include "absl/types/variant.h"
+#include "absl/strings/string_view.h"
+#include "absl/container/flat_hash_map.h"
+
+namespace wireless_android_play_analytics {
+
+class ProtoValue {
+
+ public:
+  // The only place where name can be set publicly since once set it should
+  // not be changed.
+  explicit ProtoValue(absl::string_view field_name) 
+   : field_name_(field_name) {}
+
+  virtual ~ProtoValue() = default;
+
+  static std::shared_ptr<ProtoValue> Create(absl::string_view text_proto, 
```

Comment:
I really don't think you need unique_ptr or shared_ptr here.

Try this

https://screenshot.googleplex.com/AZqx8Czq0Tc